### PR TITLE
Enable API calls for tiered index MOD-4966

### PR DESCRIPTION
**Pull request by meiravgri**
Please skip the first commit, it only fixes getting allocations overhead from `VecSimAllocator` instead of hard-coded `sizeof(size_t)`, and removes redundant module import in setup.py

### CODE REVIEW

**Review by alonre24 — COMMENTED** [review]
All looks good, with minor small comments and a suggestion to add another test with vectors that are identical, to better demonstrate the insights that we collected during the debugging phase (as discussed)

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp [hunk 1]
```diff
@@ -143,6 +150,27 @@ VecSimIndex *NewTieredIndex(const TieredHNSWParams *params,
...
 }
 
+size_t EstimateInitialSize(const TieredHNSWParams *params) {
+
+    const HNSWParams hnsw_params = params->hnswParams;
+
+    // Add size estimation of VecSimTieredIndex sub indexes.
+    size_t est = HNSWFactory::EstimateInitialSize(&hnsw_params);
+
+    BFParams bf_params = {.type = hnsw_params.type, .multi = hnsw_params.multi};
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_factory.cpp:160]: not in the initial size

File: tests/unit/test_hnsw.cpp [hunk 2]
```diff
@@ -1633,6 +1633,127 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+    size_t dim = 4;
+    size_t n = DEFAULT_BLOCK_SIZE;
+    size_t bs = DEFAULT_BLOCK_SIZE;
+    size_t efC = n;
+
+    // In this test we add identical vectors.
+    // Expected results:
+    // in level 0: each new node is connected to the M last inserted nodes, and there will be no
+    // unidirectional edges. The reason for that is that each node has room for 2 * M neighbors, but
+    // the search stops at M, so no reparations are required. for higher levels (if the max level of
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:1646]:
```suggestion
    // in level 0: each new node is connected to the M last inserted nodes, and there will be no
    // unidirectional edges. The reason for that is that each node has room for 2 * M neighbors, but
    // only connects the M vectors whose ids are the maximal (due to the secondary sorting by id in the top candidates heap). Hence, every new connection is bidirectional, so no reparations are required. for higher levels (if the max level of
```

File: tests/unit/test_hnsw.cpp [hunk 3]
```diff
@@ -1633,6 +1633,127 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+
+        size_t initial_memory = index->getAllocationSize();
+        size_t metadata_overhead_estimation = 0;
+        std::vector<size_t> nodes_per_level_hist(100, 0);
+        for (size_t i = 0; i < n; i++) {
+
+            GenerateAndAddVector<TEST_DATA_T>(index, dim, i);
+
+            size_t elem_level = hnsw_index->element_levels_[i];
+            // we generate a new incoming edges vector for each new node at each level.
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:1688]:
```suggestion
            // we expect to generate a new incoming edges vector for each new node at each level.
```

File: tests/unit/test_hnsw.cpp [hunk 4]
```diff
@@ -1633,6 +1633,127 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+                    curr_visited_at_level_hist += 1;
+                    size_t curr_idx_at_level = curr_visited_at_level_hist - 1;
+                    // The index of the vector at the current level counting backwards.
+                    size_t curr_reverse_idx_at_level =
+                        nodes_per_level_hist[level] - curr_visited_at_level_hist;
+                    auto incoming_edges = hnsw_index->getIncomingEdgesPtr(id, level);
+                    size_t incoming_edges_count = incoming_edges->size();
+
+                    // if its level 0 or there are less then M nodes at this level, none of them has
+                    // incoming edges.
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:1716]:
```suggestion
                    // If it is level 0 or there are less than M nodes at this level, none of them should have
                    // incoming edges.
```

File: tests/unit/test_hnsw.cpp [hunk 5]
```diff
@@ -1633,6 +1633,127 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+    // unidirectional edges. The reason for that is that each node has room for 2 * M neighbors, but
+    // the search stops at M, so no reparations are required. for higher levels (if the max level of
+    // the new node is higher then 0): For each level, if there are already M + 1 nodes on this
+    // level, the new node will be connected to the last M existing nodes. These M last nodes will
+    // remove one of their neighbour, so one unidirectional for each node (total of M) will be
+    // added. Eventually, we expected that each vector at this level will have M incoming edges,
+    // except the first M vectors and the last M vectors. The first M vectors will have 1, 2, ...
+    // M-1 incoming edges, according to their relative index within the level. The last M vectors
+    // will behave in a similar way, if their index is counted relatively to the last vector in the
+    // level (for example, the last vector will have 0 incoming edges)
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:1654]:
```suggestion
    // except the first M vectors and the last M vectors. The M vectors whose ids are minimal at some level should end 
    // up with 0, 1, ..., M-1 incoming edges respectively. The last M vectors will behave in a similar way (the vector with 
    // maximum id at a level should have 0 incoming edges, the one whose is the second last should have 1, etc...) 
```

**Review by alonre24 — COMMENTED** [review]
LOVED the new test :)
only some small phrasing comments about the comments

File: tests/unit/test_hnsw.cpp [hunk 6]
```diff
@@ -1633,6 +1633,127 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+                    incoming_edges->shrink_to_fit();
+                    incoming_edges_memory_overhead += incoming_edges_count * sizeof(idType);
+                }
+            }
+            // each node (except the first M and the last M) should have the same number of incoming
+            // edges.
+            if (level == 0 || nodes_per_level_hist[level] <= 2 * M) {
+                continue;
+            }
+            size_t edge_nodes = (M * (M - 1)) / 2;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_hnsw.cpp:1739]: exactly

**Review by alonre24 — APPROVED** [review]

File: tests/unit/test_hnsw.cpp [hunk 7]
```diff
@@ -1630,6 +1633,130 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+
+        // meta data per node at higher levels (level0 meta data was already allcoated at index
+        // construction)
+        size_t size_links_higher_level = sizeof(linkListSize) + M * sizeof(idType) + sizeof(void *);
+
+        size_t size_label_lookup_node = getLabelsLookupNodeSize();
+
+        size_t initial_memory = index->getAllocationSize();
+        size_t metadata_overhead_estimation = 0;
+        std::vector<size_t> nodes_per_level_hist(100, 0);
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw.cpp:1681]: represents what?

File: tests/unit/test_hnsw.cpp [hunk 8]
```diff
@@ -1630,6 +1633,130 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+    size_t n = DEFAULT_BLOCK_SIZE;
+    size_t bs = DEFAULT_BLOCK_SIZE;
+    size_t efC = n;
+
+    // In this test we add identical vectors.
+    // Expected results:
+    // in level 0: each new node is connected to the M last inserted nodes, and there will be no
+    // unidirectional edges. The reason for that is that each node has room for 2 * M neighbors, but
+    // only connects the M vectors whose ids are the maximal (due to the secondary sorting by id in
+    // the top candidates heap). Hence, every new connection is bidirectional, so no reparations are
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw.cpp:1647]: I don't think that `reparations` is the right term here

File: tests/unit/test_hnsw.cpp [hunk 9]
```diff
@@ -1630,6 +1633,130 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+                    incoming_edges_memory_overhead +=
+                        incoming_edges_count * sizeof(idType) + allocations_overhead;
+                }
+            }
+            // each node (except the first M and the last M) should have the same number of incoming
+            // edges.
+            if (level == 0 || nodes_per_level_hist[level] <= 2 * M) {
+                continue;
+            }
+            size_t edge_nodes_incoming_edges_sum = (M * (M - 1)) / 2;
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw.cpp:1743]: comments

File: tests/unit/test_hnsw.cpp [hunk 10]
```diff
@@ -1630,6 +1633,130 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
...
+    size_t dim = 4;
+    size_t n = DEFAULT_BLOCK_SIZE;
+    size_t bs = DEFAULT_BLOCK_SIZE;
+    size_t efC = n;
+
+    // In this test we add identical vectors.
+    // Expected results:
+    // in level 0: each new node is connected to the M last inserted nodes, and there will be no
+    // unidirectional edges. The reason for that is that each node has room for 2 * M neighbors, but
+    // only connects the M vectors whose ids are the maximal (due to the secondary sorting by id in
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw.cpp:1646]: Are you sure the vectors in level 0 connect only to `M` vectors? They can get up to 2M. As far as I understand, the flow is the same in each level, the only relaxtion for level 0 is that you are allowed to add up to addtional M connections without the need to "swap" outgoing edges from your earliest neighbour to the newest. This means that your observation about the increasing amount of outgoing edges for `curr_idx_at_level < M` and decreasing amount for `curr_reverse_idx_at_level < M` holds for the total number of edges (out+in) for each node in each level (meaning the other nodes will have 2M connections overall)

File: tests/unit/test_utils.h [hunk 11]
```diff
@@ -29,6 +29,20 @@ struct IndexType {
 using DataTypeSet =
     ::testing::Types<IndexType<VecSimType_FLOAT32, float>, IndexType<VecSimType_FLOAT64, double>>;
 
+// Define index type for tests that can be automatically generated for single and multi.
+template <VecSimType type, bool IsMulti, typename DataType, typename DistType = DataType>
+struct IndexTypeExtended {
+    static VecSimType get_index_type() { return type; }
+    static bool get_index_label_type() { return IsMulti; }
```
Comment (DvirDukhan) [inline tests/unit/test_utils.h:36]:
```suggestion
    static bool isMulti() { return IsMulti; }
```

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 12]
```diff
@@ -132,8 +132,8 @@ void TieredHNSWIndex<DataType, DistType>::executeInsertJobWrapper(AsyncJob *job)
     auto *insert_job = reinterpret_cast<HNSWInsertJob *>(job);
     auto *job_index = reinterpret_cast<TieredHNSWIndex<DataType, DistType> *>(insert_job->index);
     job_index->executeInsertJob(insert_job);
-    job_index->UpdateIndexMemory(job_index->memoryCtx, job_index->getAllocationSize());
     delete insert_job;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:136]: insert_job is allocated using the allocator

File: tests/unit/test_hnsw_multi.cpp [hunk 13]
```diff
@@ -836,19 +836,22 @@ TYPED_TEST(HNSWMultiTest, testSizeEstimation) {
 
     ASSERT_EQ(estimation, actual);
 
-    for (size_t i = 0; i < n; i++) {
-        GenerateAndAddVector<TEST_DATA_T>(index, dim, i % n_labels, i);
+    for (size_t i = 0; i < bs; i++) {
+        GenerateAndAddVector<TEST_DATA_T>(index, dim, bs, i);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_hnsw_multi.cpp:840]: we insert one vector per label

File: src/VecSim/algorithms/hnsw/tiered_factory.h [hunk 14]
```diff
@@ -0,0 +1,24 @@
+/*
+ *Copyright Redis Ltd. 2021 - present
+ *Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ *the Server Side Public License v1 (SSPLv1).
+ */
+#pragma once
+
+#include <cstdlib> // size_t
+#include <memory>  // std::shared_ptr
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/tiered_factory.h:9]: not needed here, also remove comments

File: src/VecSim/algorithms/hnsw/tiered_factory.cpp [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+}
+size_t EstimateInitialSize(const TieredIndexParams *params) {
+
+    size_t est = 0;
+    BFParams bf_params{};
+    if (params->primaryIndexParams->algo == VecSimAlgo_HNSWLIB) {
+        est += TieredHNSWFactory::EstimateInitialSize(params, bf_params);
+    }
+
+    est += BruteForceFactory::EstimateInitialSize(&bf_params);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/tiered_factory.cpp:68]: The brute force index itself exists as long as the tiered index exists

File: src/VecSim/index_factories/tiered_factory.h [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+namespace TieredFactory {
+VecSimIndex *NewIndex(const TieredIndexParams *params, std::shared_ptr<VecSimAllocator> allocator);
+
+// The size estimation is the sum of the buffer (brute force) and main index initial sizes
+// estimations, plus the tiered index class size. Note it does not include the size of internal
+// containers such as the job queue, as those depend on the user implementation.
+size_t EstimateInitialSize(const TieredIndexParams *params);
+size_t EstimateElementSize(const TieredIndexParams *params);
+
+}; // namespace TieredFactory
```
Comment (DvirDukhan) [inline src/VecSim/index_factories/tiered_factory.h]: empty line

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**